Character controls

JonlineRB opened the pull request:
Implemented player controls in the player script.
It uses some manual physics calculations as well as raytracing to resolve movement and collisions.
It's been adapted to the levels that can be created with the level editor.
The player object is available as a prefab.

Some changers have been introduced to demoCaller in order to have it save / load / create levels from scratch from the editor. 

To test the player, please create a level in the level editor to be used in the scene Test/Character_Test_Edited_Level. The scene has an object w/ the demoCaller script, and will load the level automatically.

Please look forward to any bugs with the player controls.

JonlineRB committed:
Auto stash before rebase of "master"

JonlineRB committed:
Auto stash before rebase of "master"

JonlineRB committed:
Auto stash before rebase of "master"

JonlineRB committed:
Auto stash before rebase of "master"

JonlineRB committed:
Early version of player controls

Bare minimum player controls.

Does not use unity's physics engine, uses raycast for collisions, has a constant run speed, as well as a limit to the fall speed

JonlineRB committed:
Early version of player controls

Bare minimum player controls.

Does not use unity's physics engine, uses raycast for collisions, has a constant run speed, as well as a limit to the fall speed

JonlineRB committed:
Early version of player controls

Bare minimum player controls.

Does not use unity's physics engine, uses raycast for collisions, has a constant run speed, as well as a limit to the fall speed

JonlineRB committed:
Early version of player controls

Bare minimum player controls.

Does not use unity's physics engine, uses raycast for collisions, has a constant run speed, as well as a limit to the fall speed

Commit:
Added basic button interaction (set block, save)

Commit:
Added Stone Block to level editor.

Commit:
Camera Edge Control

JonlineRB committed:
Merge remote-tracking branch 'origin/Character' into Character

JonlineRB committed:
Working player controls

Working player controls that use raytracing instead of unity physics.

JonlineRB committed:
Working player controls

Working player controls that use raytracing instead of unity physics.

JonlineRB committed:
Cleanup

cleanup + player prefab + player registers to gameMaster on Start()

JonlineRB committed:
Cleanup

cleanup + player prefab + player registers to gameMaster on Start()

JonlineRB committed:
Minor

saved the scene where the player script can be tested

JonlineRB committed:
Minor

saved the scene where the player script can be tested

lee17017 commented on Assets/Scripts/Input/Player.cs:
```diff
@@ -0,0 +1,217 @@
+    {
+        height = transform.localScale.y;
+        width = transform.localScale.x;
+        Gamemaster gamemaster = FindObjectOfType<Gamemaster>();
```
Use Gamemaster.Instance

JonlineRB replied:
Done.
Also moved to Awake(), like in the Level script

lee17017 commented on Assets/Scripts/Input/Player.cs:
```diff
@@ -0,0 +1,217 @@
+    public float yAccel;
+    private float yVelo = 0f;
+    private bool grounded = false;
+    private bool jumpingGrace = false;
```
What exactly does jumpingGrace mean / do?

JonlineRB replied:
Jumping grace is a period of time in which the detection of floors to land on is prevented. It's used to allow the player to jump and shortly after still be able to land.
Also see jumpingGraceTime.

lee17017 commented on Assets/Scripts/Input/Player.cs:
```diff
@@ -0,0 +1,217 @@
+            
+    }
+
+    private void FixedUpdate()
```
I think all the movement/velocity calculations should be done in FixedUpdate(including the PlayerInputHandler)

JonlineRB replied:
I will test this out.
Doing so might prevent the issue that the approximation functions have solved.

JonlineRB replied:
Done.
Did not resolve the inconsistency: where exactly is the player detected to have collided with the wall or the floor. The approximation functions still aliviate this.

lee17017 commented on Assets/Scripts/Input/Player.cs:
```diff
@@ -0,0 +1,217 @@
+        transform.position = new Vector3(transform.position.x, yValue, transform.position.z);
+    }
+
+    private void ApproximateXPosition(float direction)
```
Not sure the reason for the approximate functions but they wont work unless each block is set perfectly on whole numbers

JonlineRB replied:
The blocks are on perfectly whole nubmers though. This is to place the player in a consistent place when colliding with walls and landing. It has been buggy before.

Will try placing movement in fixed update, it might prevent the inconsistency.

JonlineRB replied:
It didn't, this is necessary.

lee17017 reviewed (changes requested):
The Player currently is still missing its tag and a rigidbody meaning it does not collide with the blocks currently.
Although I think even if you add them it wont because it needs to be kinematic (I would suggest to change the oncollisionEnter to trigger and increase the collider of the player a little)

JonlineRB committed:
Registering of player in Gamester.instance

JonlineRB committed:
Registering of player in Gamester.instance

JonlineRB committed:
Added player tag

JonlineRB committed:
Added player tag

JonlineRB committed:
Slightly increased the player collider

JonlineRB committed:
Slightly increased the player collider

MarcoTUM commented on Assets/Scripts/Level/demoCaller.cs:
```diff
@@ -4,18 +4,23 @@
-        LevelSaveLoad.Save(data, "TestLevels");
+        if(save)
+            LevelSaveLoad.Save(data, "TestLevels");
     }
```
I deleted this function in my branch and replaced it with a save button function -> merge conflict?

JonlineRB replied:
will resolve this conflict with the code from your branch

MarcoTUM commented on Assets/Scripts/Input/Player.cs:
```diff
@@ -0,0 +1,220 @@
+    public float jumpVelocity;
+
+    public float yAccel;
+    private float yVelo = 0f;
```
Maybe rename to yVelocity so have all variables uniformly named, and not some named velocity and some named velo
```suggestion
    private float yVelocity = 0f;
```

MarcoTUM commented on Assets/Scripts/Input/Player.cs:
```diff
@@ -0,0 +1,220 @@
+    public float fallVelocityLimit;
+    public float jumpVelocity;
+
+    public float yAccel;
```
```suggestion
    public float yAcceleration;
```

MarcoTUM commented on Assets/Scripts/Input/Player.cs:
```diff
@@ -0,0 +1,220 @@
+    private bool cielingCollision = false;
+    public float dropFromCielingVelocity;
+    
+    private float xVelo = 0f;
```
Maybe rename to xVelocity so have all variables uniformly named, and not some named velocity and some named velo

```suggestion
    private float xVelocity = 0f;
```

JonlineRB committed:
cleanup, collision fix

Moved movements to fixed update.
Note that approximations are still necessary.

Cleanup, variable renaming.

Modified Blocks.cs to use OnTriggerEnter, same functionality.

Updated player prefab with a trigger collider, that is slightly larger then the player, added kinematic rigidbody.

JonlineRB committed:
cleanup, collision fix

Moved movements to fixed update.
Note that approximations are still necessary.

Cleanup, variable renaming.

Modified Blocks.cs to use OnTriggerEnter, same functionality.

Updated player prefab with a trigger collider, that is slightly larger then the player, added kinematic rigidbody.

MarcoTUM committed:
Merge pull request #21 from lee17017/LevelEditorButtons

Level editor buttons

JonlineRB committed:
Revamp: no approximations

Revamp of player physics which solely relies on the ray tracing. No approximations involved!

Added the player model to the prefab for testing.

JonlineRB committed:
Revamp: no approximations

Revamp of player physics which solely relies on the ray tracing. No approximations involved!

Added the player model to the prefab for testing.

JonlineRB committed:
Character test level with level loading

JonlineRB committed:
character testing level

JonlineRB committed:
Merge remote-tracking branch 'origin/Character' into Character

lee17017 commented on Assets/Scripts/Blocks/BlockData/StoneBlock_Data.cs:
```diff
@@ -0,0 +1,8 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
```
add [System.Serializable]

JonlineRB replied:
Done

lee17017 reviewed (approved):
Clean up the testScene Folder, scene did not work for me(did my own tests)

JonlineRB committed:
Cleanup of testing scenes

JonlineRB committed:
Serializable

Added system serializable to stone block data script
